### PR TITLE
core: fix type traits

### LINKED ISSUE #7599: Incorrect element size of cv::Mat when initialised with an instance of cv::Mat_<uint32_t>.
##### System information (version)
- OpenCV => 2.4.13
- Operating System / Platform => OSX 10.10.5
- Compiler => Apple LLVM version 7.0.2 (clang-700.1.81)

##### Detailed description

The example below initialises an instance of cv::Mat with an instance of cv::Mat_<uint32_t>. Unexpectedly, the element size of the cv::Mat is different to the original cv::Mat_<uint32_t>.

The problem appears to stem from CV_ELEM_SIZE and/or CV_ELEM_SIZE1.

This problem impacts iterators returned cv::Mat_<uint32_t>::begin() and cv::Mat_<uint32_t>::end().

##### Steps to reproduce
```.cpp
#include <opencv2/core.hpp>
#include <iostream>

using namespace std;
using namespace cv;

int
main(int argc, char **argv)
{
  const size_t rows = 100;
  const size_t cols = 200;

  Mat_<uint32_t> labels(rows, cols);
  Mat generic(labels);

  cout << "labels.elemSize(): " << labels.elemSize() << endl;
  cout << "generic.elemSize(): " << generic.elemSize() << endl;

  cout << "labels.elemSize1(): " << labels.elemSize1() << endl;
  cout << "generic.elemSize1(): " << generic.elemSize1() << endl;

  auto gtype = generic.type();
  auto dtype = DataType<uint32_t>::type;
  cout << "generic.type(): " << gtype << endl;
  cout << "DataType<uint32_t>::type: " << dtype << endl;

  cout << "CV_ELEM_SIZE(DataType<uint32_t>::type): " << CV_ELEM_SIZE(DataType<uint32_t>::type) << endl;
  cout << "CV_ELEM_SIZE1(DataType<uint32_t>::type): " << CV_ELEM_SIZE1(DataType<uint32_t>::type) << endl;

  return 0;
}

// $ c++ --std=c++11 blah.cpp -I/tmp/opencv-install/include -L/tmp/opencv-install/lib -lopencv_core -lz
// $ ./a.out
// labels.elemSize(): 4
// generic.elemSize(): 8
// labels.elemSize1(): 4
// generic.elemSize1(): 8
// generic.type(): 7
// DataType<uint32_t>::type: 7
// CV_ELEM_SIZE(DataType<uint32_t>::type): 8
// CV_ELEM_SIZE1(DataType<uint32_t>::type): 8
// $
```

### PR DESCRIPTION
**Pull request by alalek**
**Merge with contrib**: https://github.com/opencv/opencv_contrib/pull/1352

Removed incorrect default values from type traits.

resolves #7599

### CODE REVIEW

File: modules/core/include/opencv2/core/traits.hpp [hunk 1]
```diff
@@ -49,11 +49,23 @@
 namespace cv
 {
 
+//#define OPENCV_TRAITS_ENABLE_DEPRECATED
+
+// Hack for CUDA modules (refactoring is required for CUDA modules)
+#ifndef OPENCV_TRAITS_ENABLE_DEPRECATED
+# include "opencv2/opencv_modules.hpp"
+# ifdef HAVE_OPENCV_CUDAFEATURES2D
+#   define OPENCV_TRAITS_ENABLE_DEPRECATED
```
Comment (mshabunin) [inline modules/core/include/opencv2/core/traits.hpp]: Is it possible to turn this on via cmake?

Comment (alalek) [reply modules/core/include/opencv2/core/traits.hpp]: via CMake we should turn on this in two places at least:
- OpenCV compilation itself
- user applications which use OpenCV

Comment (mshabunin) [reply modules/core/include/opencv2/core/traits.hpp]: @alalek , sure. Maybe something like the following lines will work in both cases:
```.cmake
if (HAVE_CUDA)
    target_compile_definitions(opencv_core PUBLIC OPENCV_TRAITS_ENABLE_DEPRECATED)
endif()
```

Comment (alalek) [reply modules/core/include/opencv2/core/traits.hpp]: Replaced

**Comment by mshabunin** [comment]
:+1:

**Pull request merged**